### PR TITLE
WI #1937 Add -ca option to LSP server

### PR DESCRIPTION
**Pull request by mayanje**

### CODE REVIEW

File: TypeCobol.LanguageServer/Workspace.cs [hunk 1]
```diff
@@ -55,6 +55,12 @@ public class Workspace
         public EventHandler<string> WarningTrigger { get; set; }
         public Queue<MessageActionWrapper> MessagesActionsQueue { get; private set; }
         private Func<string, Uri, bool> _Logger;
+        public List<string> CustomAnalyzerFiles { get; set; }
```
Comment (fm-117) [inline TypeCobol.LanguageServer/Workspace.cs:58]: The property is not used, you can remove it.

File: TypeCobol.LanguageServer/Workspace.cs [hunk 2]
```diff
@@ -250,6 +256,25 @@ public bool TryGetOpenedDocumentContext(Uri fileUri, out DocumentContext openedD
...
+            if (customAnalyzerFiles != null)
+            {
+                try
+                {
+                    _customAnalyzerProviders = customAnalyzerFiles.Select(f => AnalyzerProviderLoader.LoadProvider(f));
+                }
+                catch(Exception e)
+                {
+                    _Logger(e.Message, null);
+                }
```
Comment (fm-117) [inline TypeCobol.LanguageServer/Workspace.cs:274]: If we want to log exceptions happening during load,w e have to actually perform the load. LINQ Select is deferred so nothing happens here but on the first iteration over `_customAnalyzerProviders`

File: TypeCobol.LanguageServer/Workspace.cs [hunk 3]
```diff
@@ -55,6 +55,12 @@ public class Workspace
         public EventHandler<string> WarningTrigger { get; set; }
         public Queue<MessageActionWrapper> MessagesActionsQueue { get; private set; }
         private Func<string, Uri, bool> _Logger;
+        public List<string> CustomAnalyzerFiles { get; set; }
+        /// <summary>
+        /// Custom Analyzer Providers Loaded
+        /// </summary>
+        private IEnumerable<IAnalyzerProvider> _customAnalyzerProviders;
```
Comment (fm-117) [inline TypeCobol.LanguageServer/Workspace.cs:62]: We should store an array of `IAnalyzerProvider`, meaning an already enumerated collection, see my remark below.

**Review by fm-117 — CHANGES_REQUESTED** [review]
So you decided to keep the option as a common option ? I thought we would remove it from `GetCommonTypeCobolOptions` and then duplicate it, one specifically for CLI, the other for LS.

**Review by rooksdo — DISMISSED** [review]

File: TypeCobol.LanguageServer/Workspace.cs (hunk 1 above)
Comment (mayanje) [reply TypeCobol.LanguageServer/Workspace.cs]: done in 78ea44d

File: TypeCobol.LanguageServer/Workspace.cs (hunk 2 above)
Comment (mayanje) [reply TypeCobol.LanguageServer/Workspace.cs]: done in 78ea44d

File: TypeCobol.LanguageServer/Workspace.cs (hunk 3 above)
Comment (mayanje) [reply TypeCobol.LanguageServer/Workspace.cs]: done in 78ea44d

File: TypeCobol.LanguageServer/Workspace.cs [hunk 4]
```diff
@@ -250,6 +255,29 @@ public bool TryGetOpenedDocumentContext(Uri fileUri, out DocumentContext openedD
             }
         }
 
+        /// <summary>
+        /// Load Custom Analyzers
+        /// </summary>
+        /// <param name="customAnalyzerFiles"></param>
+        internal void LoadCustomAnalyzers(List<string> customAnalyzerFiles)
+        {
+            if (customAnalyzerFiles != null)
```
Comment (fm-117) [inline TypeCobol.LanguageServer/Workspace.cs:264]: You can replace this non-nullity test with an assert as we always instantiate the list in TypeCobolServerHost

File: TypeCobol.LanguageServer/Workspace.cs [hunk 5]
```diff
@@ -505,8 +533,13 @@ public void DidChangeConfigurationParams(string[] arguments)
 
             //Configure CFG/DFA analyzer + external analyzers if any
             var compositeAnalyzerProvider = new CompositeAnalyzerProvider();
-            compositeAnalyzerProvider.AddActivator((o, t) => CfgDfaAnalyzerFactory.CreateCfgAnalyzer(TypeCobolLanguageServer.lspcfgId, Configuration.CfgBuildingMode));
-            compositeAnalyzerProvider.AddCustomProviders(Configuration.CustomAnalyzerFiles);
+            compositeAnalyzerProvider.AddActivator((o, t) => CfgDfaAnalyzerFactory.CreateCfgAnalyzer(TypeCobolLanguageServer.lspcfgId, Configuration.CfgBuildingMode));            
+            if (this._customAnalyzerProviders != null)
```
Comment (fm-117) [inline TypeCobol.LanguageServer/Workspace.cs:537]: Same here, the array of analyzers is non-null, so an Assert is better.

**Review by fm-117 — APPROVED** [review]
See my remarks, optional changes if you want.

File: TypeCobol.LanguageServer/Workspace.cs (hunk 4 above)
Comment (mayanje) [reply TypeCobol.LanguageServer/Workspace.cs]: don in 5ec844a

File: TypeCobol.LanguageServer/Workspace.cs (hunk 5 above)
Comment (mayanje) [reply TypeCobol.LanguageServer/Workspace.cs]: don in 5ec844a

**Review by rooksdo — DISMISSED** [review]

**Review by fm-117 — APPROVED** [review]

**Review by rooksdo — APPROVED** [review]

**Pull request merged**